### PR TITLE
gh-95011: Migrate syslog module to Argument Clinic

### LINKED ISSUE #95011: Migrate syslog module to Argument Clinic
This will improve code readability and performance by delegating all of the argument parsing logic to Clinic.

### PR DESCRIPTION
**Pull request by noamcohen97**
This improves code readability and performance by delegating all of the argument parsing logic to Clinic.

Fixes #95011

* Issue: gh-95011

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2022-07-19-16-32-52.gh-issue-95011.P5t6l9.rst
Comment (arhadthedev) [inline Misc/NEWS.d/next/Core and Builtins/2022-07-19-16-32-52.gh-issue-95011.P5t6l9.rst]:
```suggestion
Migrated :mod:`syslog` to Argument Clinic. Patch by Noam Cohen.
```

**Comment by erlend-aasland** [comment]
On hold until #95007 is resolved.

File: Modules/syslogmodule.c
Comment (serhiy-storchaka) [inline Modules/syslogmodule.c]: No need to add parenthesis around "l". It was only needed because `args` was a tuple, and formatting is special for tuple.

File: Modules/syslogmodule.c
Comment (serhiy-storchaka) [inline Modules/syslogmodule.c]: Start with upper case.

File: Modules/syslogmodule.c
Comment (serhiy-storchaka) [inline Modules/syslogmodule.c]: Same as above.

**Comment by erlend-aasland** [comment]
Also, please don't force push; it does not play well with the GitHub UI. Please use `git merge --no-ff main` instead.

**Comment by arhadthedev** [comment]
Namely, it forces us to reread the whole PR from scratch instead of swiftly looking at a diff:

image

**Review by serhiy-storchaka — COMMENTED** [review]
Looking at the code I found several issues in syslog (#95041). Please wait until I fix them.

File: Modules/syslogmodule.c
Comment (erlend-aasland) [inline Modules/syslogmodule.c]:
```suggestion
```

File: Modules/syslogmodule.c
Comment (erlend-aasland) [inline Modules/syslogmodule.c]: There's a ref leak here.

File: Modules/syslogmodule.c
Comment (erlend-aasland) [inline Modules/syslogmodule.c]:
```suggestion
```

File: Modules/syslogmodule.c
Comment (erlend-aasland) [reply Modules/syslogmodule.c]: Also, this branch is not tested:

1. revert the C level changes
2. add a test for this branch (verify that it is tested using your favourite C coverage tool)
3. reapply the C level changes

Ditto for all of the audit branches. See `Lib/test/audit-tests.py` and `Lib/test/test_audit.py`.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

File: Modules/syslogmodule.c
Comment (noamcohen97) [inline Modules/syslogmodule.c:198]: `syslog_openlog_impl()` might set an exception, do you think we should call `PyErr_Clear()` in case it fails in order to swallow it?

**Comment by noamcohen97** [comment]
I have made the requested changes; please review again.
Still waiting for https://github.com/python/cpython/pull/95058 to be merged.

File: Modules/syslogmodule.c
Comment (serhiy-storchaka) [reply Modules/syslogmodule.c:198]: Good catch! Do you mind to create a separate PR for this (so it can be backported)?

I think that it is better to return NULL rather of swallowing an exception (and we can make `syslog_openlog_impl()` not swallowing *all* exceptions in future).

**Comment by serhiy-storchaka** [comment]
I have merged my fixes for known issues in #95041, but then realized that @noamcohen97 found another issue just 4 minutes before. https://github.com/python/cpython/pull/95012#discussion_r929558676 I would include the fix for it in #95058 if I see it earlier, but now we need a new PR before migrating the code to Argument Clinic. Do you mind to create it @noamcohen97? Otherwise I'll create it today at evening.

**Comment by noamcohen97** [comment]
> 

Done! https://github.com/python/cpython/pull/95264

**Comment by noamcohen97** [comment]
> I have merged my fixes for known issues in #95041, but then realized that @noamcohen97 found another issue just 4 minutes before. [#95012 (comment)](https://github.com/python/cpython/pull/95012#discussion_r929558676) I would include the fix for it in #95058 if I see it earlier, but now we need a new PR before migrating the code to Argument Clinic. Do you mind to create it @noamcohen97? Otherwise I'll create it today at evening.

Done! https://github.com/python/cpython/pull/95264

File: Modules/syslogmodule.c
Comment (noamcohen97) [reply Modules/syslogmodule.c:198]: Are you sure returning `NULL` is the right thing? As for now, if `PyTuple_New()` fails we do nothing and call `syslog(3)` anyways (https://github.com/python/cpython/blob/main/Modules/syslogmodule.c#L188).
`openlog(3)` is optional - so why skipping the entire log process?

Comment (serhiy-storchaka) [reply Modules/syslogmodule.c:198]: The function should either return NULL and set an exception, or return non-NULL and not set an exception. Violation of this rule leads to mysterious bugs in the following code. In the debug build it leads to a crash.

So you should either clear an exception, or immediately return NULL. But swallowing arbitrary exceptions is in general a bad practice -- it can silence programming or environment errors or cancel user interrupt.

`openlog(3)` is optional, as well as Python's `syslog.openlog()`, but in Python the code tries to determine ident from Python's `sys.argv[0]`, and in C it is determined from C's `argv`. They can be different. This is why the code calls `syslog_openlog()` here instead of just `openlog(3)`.

Comment (erlend-aasland) [reply Modules/syslogmodule.c:198]: Marking this as resolved, as this was fixed by #95264

**Review by erlend-aasland — APPROVED** [review]
LGTM. Good job; thanks a lot!

File: Modules/syslogmodule.c
Comment (serhiy-storchaka) [inline Modules/syslogmodule.c]: PEP 7: Restore the former formatting.

File: Modules/syslogmodule.c
Comment (serhiy-storchaka) [inline Modules/syslogmodule.c]: We can use the `str` converter here.
```suggestion
    message: str
```

File: Lib/test/test_audit.py
Comment (serhiy-storchaka) [inline Lib/test/test_audit.py]: Use `self.skipTest()` to report the test as skipped.

Or use helper `test.support.import_helper.import_module()` to check that the module is importable.

Comment (erlend-aasland) [reply Lib/test/test_audit.py]: Good point. I should do the same with the sqlite3 test further up.

File: Modules/syslogmodule.c
Comment (erlend-aasland) [reply Modules/syslogmodule.c]: Ditto for `syslog.openlog`.

Comment (serhiy-storchaka) [reply Modules/syslogmodule.c]: For `syslog.openlog` we need a Python object.

We can of course create a new `str` or `bytes` object, or allocate a raw memory, but it is simpler with the original argument.

Comment (erlend-aasland) [reply Modules/syslogmodule.c]: Yes, of course. Sorry for the noise.

**Comment by erlend-aasland** [comment]
It looks like Serhiy's comments were addressed. I'm going to land this later today.

**Comment by serhiy-storchaka** [comment]
Wait a bit more. I think that the new tests are worth to be backported, so I added them to my branch which adds other tests. I worked on this on Saturday, but I need a few more days to let it mature.

**Comment by erlend-aasland** [comment]
Ok, thanks for the heads-up. I'll mark this as don't-merge in the meantime.

**Comment by erlend-aasland** [comment]
> Wait a bit more. I think that the new tests are worth to be backported, so I added them to my branch which adds other tests. I worked on this on Saturday, but I need a few more days to let it mature.

@serhiy-storchaka: do you have an issue/PR number for this, so we can track progress?

**Comment by erlend-aasland** [comment]
> Also, please don't force push; it does not play well with the GitHub UI. Please use `git merge --no-ff main` instead.

☝🏻

**Comment by serhiy-storchaka** [comment]
Sorry for the delay. See #97953. After merging it I'll merge this PR.

**Comment by noamcohen97** [comment]
#98056 fixed the failing tests

File: Lib/test/audit-tests.py
Comment (serhiy-storchaka) [inline Lib/test/audit-tests.py]: This is a duplicate of the test above.

File: Lib/test/test_audit.py
Comment (serhiy-storchaka) [inline Lib/test/test_audit.py]: This is a duplicate of the test above.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Thank you for your contribution @noamcohen97.

**Pull request merged**
